### PR TITLE
[Perf] Avoid per-step pinned staging allocation in apply_grammar_bitmask

### LINKED ISSUE #49013: [Perf] ~2x decode throughput regression for structured outputs since #45424: apply_grammar_bitmask staging rewrite (bisected to commit, file, and hunk)
## Summary

End-to-end decode throughput for guided-JSON (xgrammar) workloads drops ~2x
between v0.23.0 and v0.24.0 (also v0.25.1). Bisected to commit
**7df3d7dada840c68b85b26b79de7f59f676d58e3** "[Core] Ensure memory is pinned
prior to async h2d copy" (#45424), then within the commit to a single file:
**vllm/v1/structured_output/utils.py — `apply_grammar_bitmask`**.

Evidence chain (identical harness, deterministic to ±1s per 2,000-request
drain; A100 80GB, Gemma-4-26B-A4B-IT bf16, max_num_seqs=64, ngram spec decode
k=5, CUDA graphs, EVERY request uses output_format json_schema → xgrammar):

| build | drain time | rate |
|---|---|---|
| 8dd1b702f (parent commit) | 513 s | 233 req/min |
| 7df3d7dad (culprit) | 985 s | 121 req/min |
| 7df3d7dad + ALL 29 v1/* files reverted to parent | 534 s | 224 req/min |
| 7df3d7dad + ONLY structured_output/utils.py reverted | 534 s | 224 req/min |
| 7df3d7dad + worker files (gpu_model_runner/gpu_input_batch) reverted | 1006 s | 119 req/min |
| 7df3d7dad + PIN_MEMORY forced False (torch_utils) | 985 s | 121 req/min |
| v0.24.0 release | 7581 s vs 3810 s on 15k corpus | 2.0x slower |

Note the PIN_MEMORY=False row: forcing the new global pin flag off does NOT
recover throughput, so the cost is in the staging rewrite itself, not merely
the pinned allocation flag.
Hunk-level split: restoring ONLY the staging block below (keeping the
commit's other change in the same function — index_tensor via
async_tensor_h2d) fully recovers baseline: **514 s / 233 req/min**. The
staging rewrite is the entire regression.

## The suspect code

#45424 changed the per-step bitmask staging in `apply_grammar_bitmask` from:

```python
sorted_bitmask = np.full(
    shape=(logits.shape[0], grammar_bitmask.shape[1]), fill_value=-1,
    dtype=grammar_bitmask.dtype,
)
...  # reorder loop fills rows
grammar_bitmask = torch.from_numpy(sorted_bitmask).to(logits.device, non_blocking=True)
```

to:

```python
sorted_bitmask_tensor = torch.full(
    (logits.shape[0], grammar_bitmask.shape[1]), -1,
    dtype=torch.from_numpy(grammar_bitmask[:0]).dtype,
    pin_memory=PIN_MEMORY,
)
sorted_bitmask = sorted_bitmask_tensor.numpy()
...  # same reorder loop, now writing into the (pinned) tensor's numpy view
grammar_bitmask = sorted_bitmask_tensor.to(logits.device, non_blocking=True)
```

For a 262k-vocab model at max_num_seqs=64 with k=5 ngram speculation this is a
~(384 x 8192) int32 = ~12.6 MB freshly allocated (normally pinned) host tensor
per engine step, plus the reorder loop writing through the numpy view of that
pinned buffer, freed again at the end of the step. This path runs every step
for every batch containing structured-output requests.

## Setup detail

- A100 80GB PCIe, torch 2.11.0 (cu13), python 3.12, offline LLM engine
  (in-process), xgrammar backend, chunked prefill 8192, TRITON_ATTN
  (forced by Gemma4 hetero head dims), async scheduling disabled (ngram).
- Symptom profile: engine steps uniformly ~2x slower at identical batch
  occupancy, acceptance rates, KV/prefix-cache hit, GPU clocks; CUDA graphs
  on/off preserves the gap; regression also present WITHOUT spec decode
  (0.25.1 no-ngram run degraded >2x as well). All requests guided-JSON.
- Release-level numbers (14,985-request production-derived corpus):
  v0.21.0 3810 s makespan / 531 accepted tok/s → v0.24.0 7581 s / 266,
  v0.25.1 7590 s / 265.

## Suggested fix

Restore the numpy staging (or keep a cached, reused pinned staging buffer
sized to (max_logits_rows, bitmask_width) instead of a fresh per-step
allocation).

Possibly related: #47047 (Gemma-4-26B-A4B on SM120/NVFP4, 0.21→0.23 slowdown
at low concurrency, unresolved — different window but same model family and
also structured-output-heavy workloads).

Full bisect logs (9 commit-level probes via per-commit wheels + 8 file/hunk-level
image probes) available on request.

### PR DESCRIPTION
**Pull request by chengcuiping**
Fixes #49013

## Purpose

Avoid the per-step pinned `torch.full(...).numpy()` staging allocation in
`apply_grammar_bitmask`. The grammar bitmask already arrives from the scheduler
as `np.ndarray[np.int32]`, so this restores NumPy host staging, preserves the
existing structured-output row reordering and speculative-token offsets, and
keeps the existing `async_tensor_h2d(out_indices, ...)` behavior.

This PR does not add a pinned ring, cache, global state, or compact mask path.
The device copy still uses `non_blocking=True`, but this PR does not claim that
pageable NumPy memory guarantees asynchronous H2D behavior.

Duplicate-work check: GitHub connector searches for `49013 grammar bitmask`,
`apply_grammar_bitmask pinned staging structured output`, related open PRs, and
related branches found no duplicate open PR or branch. The issue remains open.

AI assistance was used to prepare this change. The local `reports/issue_49013/`
benchmark and validation evidence is not included in the submitted patch.

## Test Plan

- `git diff --check`
- `.venv/bin/python -m py_compile vllm/v1/structured_output/utils.py tests/v1/structured_output/test_apply_grammar_bitmask.py`
- `.venv/bin/pre-commit run --files vllm/v1/structured_output/utils.py tests/v1/structured_output/test_apply_grammar_bitmask.py`
- `PYTHONDONTWRITEBYTECODE=1 .venv/bin/python -m pytest -q --confcutdir=tests/v1/structured_output tests/v1/structured_output/test_apply_grammar_bitmask.py -k cpu_path`
- `CUDA_VISIBLE_DEVICES=0 PYTHONDONTWRITEBYTECODE=1 .venv/bin/python -m pytest -q --confcutdir=tests/v1/structured_output tests/v1/structured_output/test_apply_grammar_bitmask.py -k cuda`
- CUDA stress: 10 repeated CUDA pytest rounds with `CUDA_VISIBLE_DEVICES=0`
- `/tmp/codex-uv-bin/uv pip check --python .venv/bin/python`

## Test Result

All final checks passed locally.

Correctness tests compare the final device bitmask/logits against the previous
torch-full staging reference for empty, all-active, reordered mixed,
speculative-offset, repeated-call, CPU, and CUDA cases. No model quality change
is expected; this is a staging-only change and correctness is exact-match.

Focused environment:

- Python 3.12.3
- torch 2.11.0+cu128, CUDA runtime 12.8
- torchaudio 2.11.0+cu128
- torchvision 0.26.0+cu128
- numpy 2.4.4
- xgrammar 0.2.3
- NVIDIA A100-SXM4-80GB, with CUDA tests run under `CUDA_VISIBLE_DEVICES=0`

Local ordinary synchronized microbenchmark, issue shape:

| case | torch_full | numpy_full | speedup | exact |
|---|---:|---:|---:|---|
| issue_384x8192_all_structured_spec5 | 3.166 ms | 2.182 ms | 1.45x | yes |
| medium_128x2048_all_structured_spec1 | 0.380 ms | 0.215 ms | 1.77x | yes |
| mixed_128x2048_every_third_spec1 | 0.151 ms | 0.135 ms | 1.12x | yes |
| mixed_384x8192_half_structured_spec5 | 1.571 ms | 1.677 ms | 0.94x | yes |
| small_32x1024_all_structured_nospec | 0.043 ms | 0.040 ms | 1.07x | yes |

Event-instrumented timing is a separate mode and should not be mixed with the
ordinary microbenchmark values. In the event-instrumented issue-shape run,
`numpy_full` total mean was 2.799 ms, with CPU staging mean 1.930 ms and H2D
event mean 0.820 ms.

The issue author's approximately 2x end-to-end recovery is from the issue
reproduction, not a local end-to-end result from this PR.

### CODE REVIEW

**Comment by njhill** [comment]
Thanks @chengcuiping. Please try https://github.com/vllm-project/vllm/pull/49919 to see if it fixes the issue.

**Comment by njhill** [comment]
@chengcuiping closing now that https://github.com/vllm-project/vllm/pull/49919 is merged. Please speak up if it didn't fix the problem for you!
